src/channel.cr

```diff
@@ -102,82 +102,136 @@ class Channel(T)
     Closed
   end
 
-  private record Sender(T), fiber : Fiber, value : T, state_ptr : DeliveryState*, select_context : SelectContext(Nil)?
-  private record Receiver(T), fiber : Fiber, value_ptr : T*, state_ptr : DeliveryState*, select_context : SelectContext(T)?
+  private struct Sender(T)
+    property fiber : Fiber
+    property data : T
+    property state : DeliveryState
+    property select_context : SelectContext(Nil)?
+    property link : Crystal::StaticList
+
+    def initialize
+      @fiber = uninitialized Fiber
+      @data = uninitialized T
+      @state = DeliveryState::None
+      @link = uninitialized Crystal::StaticList
+    end
+  end
+
+  private struct Receiver(T)
+    property fiber : Fiber
+    property data : T
+    property state : DeliveryState
+    property select_context : SelectContext(T)?
+    property link : Crystal::StaticList
+
+    def initialize
+      @fiber = uninitialized Fiber
+      @data = uninitialized T
+      @state = DeliveryState::None
+      @link = uninitialized Crystal::StaticList
+    end
+  end
 
   def initialize(@capacity = 0)
     @closed = false
-    @senders = Deque(Sender(T)).new
-    @receivers = Deque(Receiver(T)).new
+
+    @senders = uninitialized Crystal::StaticList
+    @senders.init
```

Comment:
StaticList could just have an initialize method. It would be proper to have StaticList.new (allocated onnstackz because its a struct) rather than creating an uninitialized value then initializing it, with negligible (if any) performance impact.

Reply:
Unfortunately we can't do that with simple `StaticList.new`. The `init` method will make the list become self-reference, so it requires the address of `@senders`. And  `@senders = StaticList.new` is a copy behavior. If move into initialize method, it will be initialized with incorrect address.

Reply:
Ah, I missed that. Sorry for the noise.